### Exercise 1.1

##### Create a DataFrame by reading in the `2019_Yellow_Taxi_Trip_Data.csv` file. Examine the first 5 rows.

In [8]:
import pandas as pd
data = pd.read_csv("2019_Yellow_Taxi_Trip_Data.csv")
data[1:6]


vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  \
1         1  2019-10-23T16:32:08.000  2019-10-23T16:45:26.000   
2         2  2019-10-23T16:08:44.000  2019-10-23T16:21:11.000   
3         2  2019-10-23T16:22:44.000  2019-10-23T16:43:26.000   
4         2  2019-10-23T16:45:11.000  2019-10-23T16:58:49.000   
5         2  2019-10-23T16:31:12.000  2019-10-23T16:43:20.000   

   passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
1                1           2.00           1                  N   
2                1           1.36           1                  N   
3                1           1.00           1                  N   
4                1           1.96           1                  N   
5                1           1.62           1                  N   

   pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  \
1            11            26             1         10.5    1.0      0.5   
2           163           162             1          9.5    1.0      0.5   
3           170           163             1         13.0    1.0      0.5   
4           163           236             1         10.5    1.0      0.5   
5           161            90             1          9.0    1.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
1        0.00           0.0                    0.3         12.30   
2        2.00           0.0                    0.3         15.80   
3        4.32           0.0                    0.3         21.62   
4        0.50           0.0                    0.3         15.30   
5        2.66           0.0                    0.3         15.96   

   congestion_surcharge  
1                   0.0  
2                   2.5  
3                   2.5  
4                   2.5  
5                   2.5

### Exercise 1.2

##### Find the dimensions (number of rows and number of columns) in the data.

In [10]:
data.shape


(10000, 18)

#### Exercise 1.3
##### Using the data in the `2019_Yellow_Taxi_Trip_Data.csv` file, calculate summary statistics for the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` columns.

In [12]:
summary_stats = data[['fare_amount','tip_amount','tolls_amount','total_amount']].describe()
print(summary_stats)

        fare_amount    tip_amount  tolls_amount  total_amount
count  10000.000000  10000.000000  10000.000000  10000.000000
mean      15.106313      2.634494      0.623447     22.564659
std       13.954762      3.409800      6.437507     19.209255
min      -52.000000      0.000000     -6.120000    -65.920000
25%        7.000000      0.000000      0.000000     12.375000
50%       10.000000      2.000000      0.000000     16.300000
75%       16.000000      3.250000      0.000000     22.880000
max      176.000000     43.000000    612.000000    671.800000


#### Exercise 1.4
##### Isolate the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` for the longest trip by distance (`trip_distance`).

In [17]:
max_distance_index = data['trip_distance'].idxmax()
longest_trip_info = data.loc[max_distance_index,['fare_amount','tip_amount','tolls_amount','total_amount']]
print(longest_trip_info)

fare_amount      176.0
tip_amount       18.29
tolls_amount      6.12
total_amount    201.21
Name: 8338, dtype: object


---

### Section 2

#### Exercise 2.1
##### Read in the meteorite data from the `Meteorite_Landings.csv` file, rename the `mass (g)` column to `mass`, and drop all the latitude and longitude columns. Sort the result by mass in descending order.

In [21]:
import pandas as pd
MeData = pd.read_csv('Meteorite_Landings.csv')
MeData.rename(columns={'mass(g)':'mass'},inplace=True)
MeData.drop(columns=['reclat','reclong'],inplace=True)
MeData


name     id nametype              recclass  mass (g)   fall  \
0          Aachen      1    Valid                    L5      21.0   Fell   
1          Aarhus      2    Valid                    H6     720.0   Fell   
2            Abee      6    Valid                   EH4  107000.0   Fell   
3        Acapulco     10    Valid           Acapulcoite    1914.0   Fell   
4         Achiras    370    Valid                    L6     780.0   Fell   
...           ...    ...      ...                   ...       ...    ...   
45711  Zillah 002  31356    Valid               Eucrite     172.0  Found   
45712      Zinder  30409    Valid  Pallasite, ungrouped      46.0  Found   
45713        Zlin  30410    Valid                    H4       3.3  Found   
45714   Zubkovsky  31357    Valid                    L6    2167.0  Found   
45715  Zulu Queen  30414    Valid                  L3.7     200.0  Found   

                         year             GeoLocation  
0      01/01/1880 12:00:00 AM       (50.775, 6.08333)  
1      01/01/1951 12:00:00 AM    (56.18333, 10.23333)  
2      01/01/1952 12:00:00 AM      (54.21667, -113.0)  
3      01/01/1976 12:00:00 AM       (16.88333, -99.9)  
4      01/01/1902 12:00:00 AM     (-33.16667, -64.95)  
...                       ...                     ...  
45711  01/01/1990 12:00:00 AM       (29.037, 17.0185)  
45712  01/01/1999 12:00:00 AM     (13.78333, 8.96667)  
45713  01/01/1939 12:00:00 AM       (49.25, 17.66667)  
45714  01/01/2003 12:00:00 AM     (49.78917, 41.5046)  
45715  01/01/1976 12:00:00 AM  (33.98333, -115.68333)  

[45716 rows x 8 columns]

#### Exercise 2.2
##### Using the meteorite data from the `Meteorite_Landings.csv` file, update the `year` column to only contain the year, convert it to a numeric data type, and create a new column indicating whether the meteorite was observed falling before 1970. Set the index to the `id` column and extract all the rows with IDs between 10,036 and 10,040 (inclusive) with `loc[]`.

###### **Hint 1**: Use `year.str.slice()` to grab a substring.

###### **Hint 2**: Make sure to sort the index before using `loc[]` to select the range.

###### **Bonus**: There's a data entry error in the `year` column. Can you find it? (Don't spend too much time on this.)

#### Exercise 2.3
##### Using the meteorite data from the `Meteorite_Landings.csv` file, create a pivot table that shows both the number of meteorites and the 95th percentile of meteorite mass for those that were found versus observed falling per year from 2005 through 2009 (inclusive). Hint: Be sure to convert the `year` column to a number as we did in the previous exercise.

#### Exercise 2.4
##### Using the meteorite data from the `Meteorite_Landings.csv` file, compare summary statistics of the mass column for the meteorites that were found versus observed falling.

#### Exercise 2.5
##### Using the taxi trip data in the `2019_Yellow_Taxi_Trip_Data.csv` file, resample the data to an hourly frequency based on the dropoff time. Calculate the total `trip_distance`, `fare_amount`, `tolls_amount`, and `tip_amount`, then find the 5 hours with the most tips.

---

### Section 3

#### Exercise 3.1
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create box plots for traveler throughput for each year in the data. Hint: Pass `kind='box'` into the `plot()` method to generate box plots.

#### Exercise 3.2
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create a heatmap that shows the 2019 TSA median traveler throughput by day of week and month.

#### Exercise 3.3
##### Annotate the medians in the box plot from *[Exercise 3.1](#Exercise-3.1)*. Hint: The `x` coordinates will be 1, 2, and 3 for 2019, 2020, and 2021, respectively. Alternatively, to avoid hardcoding values, you can use the `Axes.get_xticklabels()` method, in which case you should look at the [documentation](https://matplotlib.org/stable/api/text_api.html) for the `Text` class.